### PR TITLE
[VoiceOver] Update Guidelines Doc

### PR DESCRIPTION
**Pull request by shiki**
WIP. View the updated doc [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/update/voiceover-doc/docs/voiceover-guidelines.md).

### CODE REVIEW

File: docs/voiceover-guidelines.md [hunk 1]
```diff
@@ -1,61 +1,179 @@
...
-- Not include the type of control or view
-- Begin with a capitalized word
+
+- Describe the element briefly.
+- Not include the type of control or view.
+- Begin with a capitalized word.
 - Not end with a period.
 - Be localized.
 
 Beginning with a capitalized word and not ending with a period helps VoiceOver read the label with the appropiate inflection.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
Beginning with a capitalized word and not ending with a period helps VoiceOver read the label with the appropriate inflection.
```

File: docs/voiceover-guidelines.md [hunk 2]
```diff
@@ -1,61 +1,179 @@
...
+
+- Prefer to place the most important elements first. The VoiceOver user can prefer to skip if they've already listened to what they need. This is why we placed the `excerpt` last in the example.
+- Don't forget the periods when concatenating. They make VoiceOver pause which helps in comprehension.
+- The `"Excerpt".` static text in the example is used as a separator and _signals_ that a very long text will be read by VoiceOver.
+
+### <a name="navigation-order"></a>Navigation Order
+
+VoiceOver may not navigate views in the order that you'd naturally expect. For example, this `UITableViewCell` has nested stack views where:
+
+- The vertical stack view is the parent and multiple horizontal stack views are children.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
- The vertical stack view is the parent, and multiple horizontal stack views are children. 
```

File: docs/voiceover-guidelines.md (hunk 1 above)
Comment (shiki) [reply docs/voiceover-guidelines.md]: Oof. Thanks! 😄

File: docs/voiceover-guidelines.md [hunk 3]
```diff
@@ -1,61 +1,179 @@
...
-![](images/voiceover-guidelines.png)
+With this, VoiceOver will read the label (e.g. `"First name"`) and the value (e.g. `"John"`).
 
-The reading order is expected to be spoken as, "Subtotal: nine hundred ninety-nine dollars and ninety nine cents". In this case, VoiceOver defaults to reading the first item in each horizontal stack view, followed by the second. The default order speaks a list of titles: "Subtotal - Discount - Shipping - Taxes" and so on, followed by the values: "$999.99, -$601.00, $0.01, $333.33", making the information difficult to comprehend.
+### <a name="grouping-elements"></a>Grouping Elements
 
-Use `.accessibilityElements` on the parent view to list the desired reading order of the subviews.
+If a group of elements represent a single unit of information, consider grouping them into one accessibility element. This helps reduce clutter and makes your app easier to understand and navigate. 
+
+Take the following custom `UITableViewCell` as an example. It has at least 5 accessible elements.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
Take the following custom `UITableViewCell` as an example. It has at least five accessible elements. 
```

File: docs/voiceover-guidelines.md [hunk 4]
```diff
@@ -1,61 +1,179 @@
...
-Use `.accessibilityElements` on the parent view to list the desired reading order of the subviews.
+If a group of elements represent a single unit of information, consider grouping them into one accessibility element. This helps reduce clutter and makes your app easier to understand and navigate. 
+
+Take the following custom `UITableViewCell` as an example. It has at least 5 accessible elements. 
+
+<img src="images/voiceover-guidelines/group-elements-before.png" width="320">
+
+Since there are potentially more cells like this in the table, it would be very easy for a VoiceOver user to lose context. To improve this, you can:
+
+- Group the elements together by concatenating the information in the `UITableViewCell`'s `accessibilityLabel`.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
- Group the elements by concatenating the information in the `UITableViewCell`'s `accessibilityLabel`.
```
The "together" here isn't necessary since it creates a tautology 😉

File: docs/voiceover-guidelines.md [hunk 5]
```diff
@@ -1,61 +1,179 @@
...
+When the cell is focused, VoiceOver would then speak something like this:
+
+```
+Post by Carolyn Wells, from Discover. 6 days ago. Following. 
+Around the World with WordPress: Jamaica. 
+Excerpt. Today, we’re highlighting five sites from the island paradise of Jamaica.
+```
+
+- Prefer to place the most important elements first. The VoiceOver user can prefer to skip if they've already listened to what they need. This is why we placed the `excerpt` last in the example.
+- Don't forget the periods when concatenating. They make VoiceOver pause which helps in comprehension.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
- Don't forget the periods when concatenating. They make VoiceOver pause, which helps in comprehension.
```

Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
- Don't forget the periods when concatenating. They make VoiceOver pause which helps to prevent incomprehension.
```
Since incomprehension is a word, "in comprehension" might be misrepresented.

File: docs/voiceover-guidelines.md [hunk 6]
```diff
@@ -65,7 +183,220 @@ contentView.accessibilityElements = [subtotalTitleLabel,
...
+
+Once the element disappears, you should send another notification but with a `nil` argument. This makes VoiceOver immediately move the focus to a different element on the screen.
+
+```swift
+UIAccessibility.post(notification: .layoutChanged, argument: nil)
+```
+
+### <a name="prefer-disabling"></a>Prefer Disabling Instead of Hiding Elements
+
+A common UI pattern is hiding elements, such as buttons, until users are able to use them.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
A common UI pattern is hiding elements, such as buttons, until users can use them. 
```
Simplifying :)

File: docs/voiceover-guidelines.md (hunk 4 above)
Comment (shiki) [reply docs/voiceover-guidelines.md]: TIL **tautology** 😄 🙇

File: docs/voiceover-guidelines.md [hunk 7]
```diff
@@ -65,7 +183,220 @@ contentView.accessibilityElements = [subtotalTitleLabel,
...
+
+### <a name="toggle-buttons"></a>Use Static Labels for Toggle Buttons
+
+In general, avoid changing the `accessibilityLabel` of elements after it's already been set. If an element's `accessibilityLabel` changes over time, users may think that it is **a different element**. This can be confusing, especially if the element is at the same position on the screen.
+
+If the behavior or meaning of an element, such as a toggle button, changes over time, consider updating the [`accessibilityValue`](https://developer.apple.com/documentation/objectivec/nsobject/1615117-accessibilityvalue) instead. The `accessibilityValue` is read by VoiceOver right after the `accessibilityLabel`. This keeps the user informed that they have the same element focused but it has a different state, and will have a different behavior when activated.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]: This sentence is a bit wordy. It would be good if it could be reduced a bit.

File: docs/voiceover-guidelines.md [hunk 8]
```diff
@@ -65,7 +183,220 @@ contentView.accessibilityElements = [subtotalTitleLabel,
...
+mostPopularInsightCell.manageButton.accessibilityLabel = "Manage Most Popular Insight"
+```
+
+With this, a user focusing a `manageButton` would have enough context. They wouldn't need to go navigate back to the cell's title to deduce the context themselves.
+
+### <a name="escape-gesture"></a>Support Escape Gesture for Custom Modal Views
+
+<img src="images/voiceover-guidelines/custom-dialog.gif" width="240">
+
+VoiceOver has a standard escape gesture which dismisses an alert or returns to the previous screen. You can perform the gesture by moving two fingers back and forth three times quickly, making a “z”.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
VoiceOver has a standard escape gesture that dismisses an alert or returns to the previous screen. You can perform the gesture by moving two fingers back and forth three times quickly, making a “z”. 
```

File: docs/voiceover-guidelines.md [hunk 9]
```diff
@@ -65,7 +183,220 @@ contentView.accessibilityElements = [subtotalTitleLabel,
...
+
+### <a name="escape-gesture"></a>Support Escape Gesture for Custom Modal Views
+
+<img src="images/voiceover-guidelines/custom-dialog.gif" width="240">
+
+VoiceOver has a standard escape gesture which dismisses an alert or returns to the previous screen. You can perform the gesture by moving two fingers back and forth three times quickly, making a “z”. 
+
+This is generally supported by the all the UIKit components. If you make custom (modal) views, however, VoiceOver may not know how to dismiss it. This can potentially lock your user within that view forever.
+
+You can support the gesture by overridding [`accessibilityPerformEscape`](https://developer.apple.com/documentation/objectivec/nsobject/1615091-accessibilityperformescape) in your custom `UIView` or `UIViewController`.
```
Comment (jd-alexander) [inline docs/voiceover-guidelines.md]:
```suggestion
You can support the gesture by overriding [`accessibilityPerformEscape`](https://developer.apple.com/documentation/objectivec/nsobject/1615091-accessibilityperformescape) in your custom `UIView` or `UIViewController`. 
```

File: docs/voiceover-guidelines.md (hunk 7 above)
Comment (shiki) [reply docs/voiceover-guidelines.md]: I shall try. Thank you for pointing it out. 😬
